central/scannerdefinitions/handler/handler.go

```diff
@@ -388,81 +400,103 @@ func (h *httpHandler) openOfflineBlob(ctx context.Context, blobName string) (*vu
+			utils.IgnoreError(online.Close)
+		}
+	}()
+
 	switch t {
 	case v2UpdaterType:
 		if opts.fileName == "" {
-			return &vulDefFile{File: openedFile, modTime: onlineTime}, nil
+			break
 		}
+		// We want to find a specific file inside the toplevel online file.
 		fallthrough
 	case mappingUpdaterType:
-		// openFromArchive will copy the contents of opts.fileName from openedFile to
-		// targetFile. Because of this, openedFile is not needed outside this function,
-		// so close it here.
-		defer utils.IgnoreError(openedFile.Close)
-		targetFile, _, err := h.openFromArchive(openedFile.Name(), opts.fileName)
+		targetFile, _, err := h.openFromArchive(online.Name(), opts.fileName)
 		if err != nil {
 			return nil, err
 		}
-		return &vulDefFile{File: targetFile, modTime: onlineTime}, nil
+		f.File = targetFile
 	case vulnerabilityUpdaterType:
-		return &vulDefFile{File: openedFile, modTime: onlineTime}, nil
+		// Always return the toplevel online file.
+	default:
+		return nil, fmt.Errorf("unknown updater type: %s", t)
 	}
-	return nil, fmt.Errorf("unknown Scanner V4 updater type: %s", t)
+
+	success = true
+	return f, nil
 }
 
 // getUpdater gets or creates an updater for the scanner definitions identified
 // by the given updater type and a URL path to the definitions file. If the
 // updater was created, it is no started here, callers are expected to start it.
-func (h *httpHandler) getUpdater(t updaterType, urlPath string) *requestedUpdater {
+func (h *httpHandler) getUpdater(t updaterType, urlPath string) (*requestedUpdater, error) {
 	h.updatersLock.Lock()
 	defer h.updatersLock.Unlock()
 
 	fileName := strings.ReplaceAll(filepath.Join(t.String(), urlPath), "/", "-")
-	updater, exists := h.updaters[fileName]
+	u, exists := h.updaters[fileName]
 	if !exists {
 		var updateURL *url.URL
-		var ext string
 		switch t {
 		case mappingUpdaterType:
 			updateURL = scannerUpdateBaseURL.JoinPath(scannerV4MappingSubDir, scannerV4MappingFile)
-			ext = ".zip"
 		case vulnerabilityUpdaterType:
 			updateURL = scannerUpdateBaseURL.JoinPath(scannerV4VulnSubDir, urlPath)
-			ext = ".json.zst"
-		default: // uuid
+		case v2UpdaterType:
 			updateURL = scannerUpdateBaseURL.JoinPath(urlPath, scannerV2DiffFile)
-			ext = ".zip"
+		default:
+			return nil, fmt.Errorf("unknown updater type: %s", t)
 		}
 		filePath := filepath.Join(h.dataDir, fileName)
-		// Use a default extension if the URL path does not contain one.
-		if filepath.Ext(fileName) == "" {
-			filePath += ext
-		}
-		updater = &requestedUpdater{
+		u = &requestedUpdater{
 			updater: newUpdater(file.New(filePath), client, updateURL.String(), h.updaterInterval),
 		}
-		h.updaters[fileName] = updater
+		h.updaters[fileName] = u
 	}
 
-	updater.lastRequestedTime = time.Now()
-	return updater
+	u.lastRequestedTime = time.Now()
+	return u, nil
 }
 
 func (h *httpHandler) post(w http.ResponseWriter, r *http.Request) {
+	if r.Body == nil {
+		writeErrorBadPOSTRequest(w)
+		return
+	}
```

Comment:
No one else is using `writeErrorBadPOSTRequest()` and the error message seems specific to this particular case, let's just hard code the error logic here instead of a separate function?

Reply:
sounds good to me